README.md
# `citerelations`


## What it is

`citerelations` is a cross-platform library for working with sets of triples describing relations among citable objects.

## Current version: 2.7.1

Status:  in active development.  See [release notes](releases.md)

## License

[GPL 3.0](https://opensource.org/licenses/gpl-3.0.html)


## Documentation

See <https://cite-architecture.github.io/citerelations/>.

## Using, building, testing

`citerelations` is compiled for both the JVM and ScalaJS using scala versions 2.10, 2.11 and 2.12.

Binaries for all three versions are available from the Nexus repository on <terracotta.hpcc.uh.edu/nexus>.

If you are using sbt, include `Resolver.jcenterRepo` in your list of resolvers

```scala
	resolvers += "Nexus" at "https://terracotta.hpcc.uh.edu/nexus/repository/maven-releases/",
```

and add this to your library dependencies:

    "edu.holycross.shot" %%% "citerelations" % VERSION

For maven, ivy or gradle equivalents, refer to https://bintray.com/neelsmith/maven/citerelations.

To build from source and test, use normal sbt commands (`compile`, `test` ...).


releases.md
# `citerelations`: release notes

**2.7.1**: Artifact published now at https://terracotta.hpcc.uh.edu/nexus/repository/maven-releases/

**2.7.0**: Add support for SJS 1.0.0.

**2.6.0**:  Add logging with airframe library.

**2.5.2**:  Really update `xcite` library.


**2.5.1**:  Update `xcite` library.

**2.5.0**  Add `cex` functions on `CiteTriple` and `CiteRelationSet`.

**2.4.1**: Bug fixes in `xcite` library.


**2.4.0**:  New functionality in updated `xcite` library.

**2.3.0**:  New functionality in updated `xcite` library.

**2.2.0**: New functionality in updated `xcite` library.

**2.1.1**:  Modest change to build configuration.

**2.1.0**:  Added `[CiteRelationSet].verbs`, returning a `Set[CiteUrn]` listing the verbs represented in the CiteRelationSet.

**2.0.4**:  Fixed Issue #6.

**2.0.3**:  Bug fix in update to `cex` parser.

**2.0.2**:  Incorporate bug fixes in library dependencies.


**2.0.1**:  Incorporate bug fixes in library dependencies.

**2.0.0**:  API-breaking upgrade to support CEX 3.0.

**1.1.1**:  Fixes a bug in handling empty data blocks.

**1.1.0**: Support full CEX format in constructing `CiteRelationSet`.

**1.0.2**: Correct group used for cross-publication.

**1.0.1**: Experiment with cross-publication settings.

**1.0.0**:  Initial release.


shared/src/test/scala/citerelations/CiteTripleSpec.scala
package edu.holycross.shot.citerelation
import org.scalatest.FlatSpec
import edu.holycross.shot.cite._
import scala.io.Source


class CiteTripleSpec extends FlatSpec {


  val venetusA = CtsUrn("urn:cts:greekLit:tlg0012.tlg001.msA:")
  val twelveverso = Cite2Urn("urn:cite2:hmt:msA.r1:12v")


  "The CiteTriple object" should "form a triple from a single line of delimited text" in {
    val cex = "urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v"
    val trip = CiteTriple(cex)
    trip match {
      case ct: CiteTriple => {
        assert(ct.urn1 == CtsUrn("urn:cts:greekLit:tlg0012.tlg001.msA:2.1"))
        assert(ct.urn2 == Cite2Urn("urn:cite2:hmt:msA.r1:12v"))
        assert(ct.relation == Cite2Urn("urn:cite2:cite:dseverbs.r1:appearsOn"))
      }
      case _ => fail("Should have created a CiteTriple")
    }
  }

  "A CITE triple" should "accept a CTS URN for its first relation" in {
    val il1_1 = CtsUrn("urn:cts:greekLit:tlg0012.tlg001.msA:1.1")
    val img = Cite2Urn("urn:cite2:hmt:vaimg.r1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901")
    val appearsOn = Cite2Urn("urn:cite2:hmt:verbs:appearsOn")

    val triple = CiteTriple(il1_1,appearsOn, img)

    triple.urn1 match {
      case cts: CtsUrn => assert(true)
      case _ => fail("Should have found a CtsUrn")
    }
  }

  it should "compare a specified CTS URN to its first URN" in {
      val cex = "urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v"
      val triple = CiteTriple(cex)

      assert(triple.ctsUrn1Match(venetusA))
  }

  it should "compare a specified CTS URN to its second URN" in {
    val cex = "urn:cite2:hmt:msA.r1:12v#urn:cite2:cite:dseverbs.r1:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:2.1"
    val triple = CiteTriple(cex)
    assert(triple.ctsUrn2Match(venetusA))
  }


  it should "compare any type of URN to its first URN" in {
    val cex = "urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v"
    val triple = CiteTriple(cex)

    assert(triple.urn1Match(venetusA))
    assert(triple.urn1Match(twelveverso) == false)
  }


  it should "compare any type of URN to its second URN" in {
    val cex = "urn:cite2:hmt:msA.r1:12v#urn:cite2:cite:dseverbs.r1:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:2.1"
    val triple = CiteTriple(cex)
    assert(triple.urn2Match(venetusA))
    assert(triple.urn2Match(twelveverso) == false)
  }


  it should "compare an identified CITE2 URN to its first URN" in {
    val cex = "urn:cite2:hmt:msA.r1:12v#urn:cite2:cite:dseverbs.r1:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:2.1"
    val triple = CiteTriple(cex)
    assert(triple.citeUrn1Match(twelveverso))
    //assert(triple.citeUrn1Match(venetusA) == false)

  }

  it should "compare an identified CITE2 URN to its second URN" in {
    val cex = "urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v"
    val triple = CiteTriple(cex)
    assert(triple.citeUrn2Match(twelveverso))
    //assert(triple.citeUrn1Match(venetusA) == false)

  }

  it should "compare an unidentified CITE2 URN to its first URN" in {
    val cex = "urn:cite2:hmt:msA.r1:12v#urn:cite2:cite:dseverbs.r1:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:2.1"
    val triple = CiteTriple(cex)
    assert(triple.urn1Match(twelveverso))
    assert(triple.urn1Match(venetusA) == false)
  }

  it should "compare an unidentified CITE2 URN to its second URN" in {
    val cex = "urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v"
    val triple = CiteTriple(cex)
    assert(triple.urn2Match(twelveverso))
    assert(triple.urn2Match(venetusA) == false)

  }

  it should "generate a CEX string" in {
    val srcCex = "urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v"
    val triple = CiteTriple(srcCex)
    assert(triple.cex() == srcCex)
  }
}


shared/src/test/scala/citerelations/DseRelationSpec.scala
package edu.holycross.shot.citerelation
import org.scalatest.FlatSpec
import edu.holycross.shot.cite._
import scala.io.Source

/**
*/
class DseRelationSpec extends FlatSpec {


  "A CITE relation set" should "parse the standard DSE test data set" in {
    val dseExample = """
#!relations
// 1. Relation of text passages to text-bearing surface:
urn:cts:greekLit:tlg0012.tlg001.msA:1.1#urn:cite2:cite:dseverbs.2017a:appearsOn#urn:cite2:hmt:msA.2017a:12r
urn:cite2:hmt:msA.2017a:12r#urn:cite2:cite:dseverbs.2017a:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:1.1
// 2. Relation of text passages to documentary image:
urn:cts:greekLit:tlg0012.tlg001.msA:1.1#urn:cite2:cite:dseverbs.2017a:illustratedBy#urn:cite2:hmt:vaimg.v1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901
urn:cite2:hmt:vaimg.v1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901#urn:cite2:cite:dseverbs.2017a:illustrates#urn:cts:greekLit:tlg0012.tlg001.msA:1.1
// 3. Relation of text-bearing surface to documentary image:
urn:cite2:hmt:msA.2017a:12r#urn:cite2:cite:dseverbs.2017a:illustratedBy#urn:cite2:hmt:vaimg.2017a:VA012RN_0013
urn:cite2:hmt:vaimg.2017a:VA012RN_0013#urn:cite2:cite:dseverbs.2017a:illustrates#urn:cite2:hmt:msA.2017a:12r
"""

    val relations = CiteRelationSet(dseExample)

  }



}


shared/src/test/scala/citerelations/CiteRelationSpec.scala
package edu.holycross.shot.citerelation
import org.scalatest.FlatSpec
import edu.holycross.shot.cite._
import scala.io.Source

/**
*/
class CiteRelationSpec extends FlatSpec {

val cexSrc = """
#!relations

//  SVO triplets :

// Iliad 1.1
urn:cts:greekLit:tlg0012.tlg001.msA:1.1#urn:cite2:cite:dseverbs.r1:illustratedBy#urn:cite2:hmt:vaimg.r1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901
urn:cite2:hmt:vaimg.r1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901#urn:cite2:cite:dseverbs.r1:illustrates#urn:cts:greekLit:tlg0012.tlg001.msA:1.1

// Iliad 2.1
urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:illustratedBy#urn:cite2:hmt:vaimg.r1:VA012VN_0514@0.4956,0.2191,0.3254,0.02158
urn:cite2:hmt:vaimg.r1:VA012VN_0514@0.4956,0.2191,0.3254,0.02158#urn:cite2:cite:dseverbs.r1:illustrates#urn:cts:greekLit:tlg0012.tlg001.msA:2.1

// MS 12r
urn:cite2:hmt:msA.r1:12r#urn:cite2:cite:dseverbs.r1:illustratedBy#urn:cite2:hmt:vaimg.r1:VA012RN_0013
urn:cite2:hmt:vaimg.r1:VA012RN_0013#urn:cite2:cite:dseverbs.r1:illustrates#urn:cite2:hmt:msA.r1:12r

// MS 12v
urn:cite2:hmt:msA.r1:12v#urn:cite2:cite:dseverbs.r1:illustratedBy#urn:cite2:hmt:vaimg.r1:VA012VN_0514
urn:cite2:hmt:vaimg.r1:VA012VN_0514#urn:cite2:cite:dseverbs.r1:illustrates#urn:cite2:hmt:msA.r1:12v

// Iliad 1.1 + MS 12r
urn:cts:greekLit:tlg0012.tlg001.msA:1.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12r
urn:cite2:hmt:msA.r1:12r#urn:cite2:cite:dseverbs.r1:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:1.1

// Iliad 2.1 + MS 12v
urn:cts:greekLit:tlg0012.tlg001.msA:2.1#urn:cite2:cite:dseverbs.r1:appearsOn#urn:cite2:hmt:msA.r1:12v
urn:cite2:hmt:msA.r1:12v#urn:cite2:cite:dseverbs.r1:hasOnIt#urn:cts:greekLit:tlg0012.tlg001.msA:2.1
"""

  val relations = CiteRelationSet(cexSrc)


  // Query on notional versions of same objects:
  val msAiliad = CtsUrn("urn:cts:greekLit:tlg0012.tlg001.msA:")
  val matchPage = Cite2Urn("urn:cite2:hmt:msA:12r")
  val matchImg = Cite2Urn("urn:cite2:hmt:vaimg:VA012RN_0013")

  val dseverbs = Cite2Urn("urn:cite2:cite:dseverbs:")
  val hasonit = Cite2Urn("urn:cite2:cite:dseverbs.r1:hasOnIt")

  "A CITE relation set" should "offer a constructor signature for instantiating a corpus from a CEX text source" in  {

    relations match {
      // 3 symmetrical pairs:
      case crset: CiteRelationSet => assert(crset.size == 12)
      case _ => fail("Should have created CiteRelationSet")
    }
  }


  it should "filter a relation set by an unspecified CTS URN on URN 1" in {
    val filtered = relations.urn1Match(msAiliad)
    assert(filtered.size == 4)
  }
  it should "filter a relation set by CITE2 URN on URN 1" in  {
    val filteredPage = relations.urn1Match(matchPage)
    assert(filteredPage.size == 2)

    val filteredImage = relations.urn1Match(matchImg)
    assert(filteredImage.size == 2)
  }


  it should "filter a relation set by CTS URN on URN 2" in {
    val filtered = relations.urn2Match(msAiliad)
    assert(filtered.size == 4)
  }
  it should "filter a relation set by CITE2 URN on URN 2" in {

    val filteredPage = relations.urn2Match(matchPage)
    assert(filteredPage.size == 2)

    val filteredImage = relations.urn2Match(matchImg)
    assert(filteredImage.size == 2)
  }


  it should "filter a relation set by CTS URN on either URN" in {
    val filtered = relations ~~ msAiliad
    assert(filtered.size == 8)

  }
  it should "filter a relation set by CITE2 URN on either URN" in {
    val filteredPage = relations ~~ matchPage
    assert(filteredPage.size == 4)

    val filteredImage = relations ~~ matchImg
    assert(filteredImage.size == 4)
  }

  it should "fiter a relation set by verb URN" in {
    val filtered = relations.verb(dseverbs)
    assert(filtered.size == 12)
  }

  it should "fiter a relation set by specific verb URN" in {
    val filtered = relations.verb(hasonit)
    assert(filtered.size == 2)
  }

  it should "serialize to CEX" in {
    val cex = relations.cex()
    val expected = cexSrc.split("\n").toVector.filter(_.nonEmpty).filterNot(_.startsWith("//")).mkString("\n")
    println("\n\nActual: " + cex.size)
    println(cex)
    println("\n\nExpeted: " + expected.size)
    println(expected)
    val cexV = cex.split("\n").toVector.sorted
    val expV = expected.split("\n").toVector.sorted
    assert (cexV == expV)

  }


}


shared/src/main/scala/citerelations/CiteTriple.scala
package edu.holycross.shot.citerelation

import edu.holycross.shot.cite._


import scala.io.Source
import scala.collection.mutable.ArrayBuffer


import wvlet.log._
import wvlet.log.LogFormatter.SourceCodeLogFormatter

import scala.scalajs.js
import scala.scalajs.js.annotation._


@JSExportAll case class CiteTriple(urn1: Urn, relation: Cite2Urn, urn2: Urn) {


  def cex(delimiter: String = "#"): String = {
    s"${urn1}${delimiter}${relation}${delimiter}${urn2}"
  }

  /** True if "subject" matches a given CTS URN.
  *
  * @param urn A CtsUrn to compare to the subject of this relation.
  */
  def ctsUrn1Match(urn: CtsUrn): Boolean = {
    urn1 match {
      case cts: CtsUrn => cts ~~ urn
      case _ => false
    }
  }


  /** True if "object" matches a given CTS URN.
  *
  * @param urn A CtsUrn to compare to the subject of this relation.
  */
  def ctsUrn2Match(urn: CtsUrn): Boolean = {
    urn2 match {
      case cts: CtsUrn => cts ~~ urn
      case _ => false
    }
  }

  /** True if "subject" matches a given CITE2 URN.
  *
  * @param urn A Cite2Urn to compare to the subject of this relation.
  */
  def citeUrn1Match(urn: Cite2Urn): Boolean = {
    urn1 match {
      case cite2: Cite2Urn => cite2 ~~ urn
      case _ => false
    }
  }

  /** True if "object" matches a given CITE2 URN.
  *
  * @param urn A Cite2Urn to compare to the subject of this relation.
  */
  def citeUrn2Match(urn: Cite2Urn): Boolean = {
    urn2 match {
      case cite2: Cite2Urn => cite2 ~~ urn
      case _ => false
    }
  }


  /** True if "subject" matches a given URN (CTS or CITE2 URN).
  *
  * @param urn URN to match.
  */
  def urn1Match (urn: Urn): Boolean = {
    urn match {
      case cts: CtsUrn => {
        ctsUrn1Match(cts)
      }
      case c2: Cite2Urn =>  {
        citeUrn1Match(c2)
      }
    }
  }


  /** True if "object" matches a given URN (CTS or CITE2 URN).
  *
  * @param urn URN to match.
  */
  def urn2Match (urn: Urn): Boolean = {
     urn match {
      case cts: CtsUrn => {
        ctsUrn2Match(cts)
      }
      case c2: Cite2Urn =>  {
        citeUrn2Match(c2)
      }
    }
  }


  /** True if "verb" URN matches a
  * a given URN.
  *
  * @param urn URN to match.
  */
  def verb (verb: Cite2Urn): Boolean = {
    relation ~~ verb
  }

  /** True if either "subject" or "object" URN match a
  * a given URN.
  *
  * @param urn URN to match.
  */
  def ~~(urn: Urn): Boolean = {
    (urn1Match(urn) || urn2Match(urn))
  }

}


/** Factory object for creating [[CiteTriple]]s from CEX source.
*/
object CiteTriple extends LogSupport {

  /** Create a CiteTriple from a single line of delimited text.
  *
  * @param cex CEX description of a single CITE relation.
  * @param delimiter Delimiting string separating S-V-O.
  */
  def apply(cex: String, delimiter: String = "#"): CiteTriple = {
    val columns = cex.split(delimiter)
    columns.size match {
      case 3 =>   CiteTriple(CiteRelationSet.urnFromString(columns(0)), Cite2Urn(columns(1)), CiteRelationSet.urnFromString(columns(2)))
      case _ => {
        val msg = s"Wrong number of columns in ${cex}"
        warn(msg)
        throw new CiteRelationException(msg)
      }
    }

  }
}


shared/src/main/scala/citerelations/CiteRelationSet.scala
package edu.holycross.shot.citerelation

import edu.holycross.shot.cite._
import edu.holycross.shot.cex._


import scala.io.Source
import scala.collection.mutable.ArrayBuffer


import wvlet.log._
import wvlet.log.LogFormatter.SourceCodeLogFormatter



import scala.scalajs.js
import scala.scalajs.js.annotation._


@JSExportAll case class CiteRelationSet (relations: Set[CiteTriple]) extends LogSupport {

  /** True if relations set is empty.*/
  def isEmpty: Boolean = {
    (relations.size == 0)
  }

  /** Filter set for verb URN matching a given URN.
  *
  * @param u CITE2 URN to match.
  */
  def verb(verbUrn: Cite2Urn): CiteRelationSet = {
    val matchingRelations = relations.filter(_.verb(verbUrn))
      CiteRelationSet(matchingRelations)
  }

  /** Find number of relations in set.
  */
  def size: Int = {
    relations.size
  }

  /** Return a Set of CiteUrns representing all verbs
  *
  */
  def verbs:Set[Cite2Urn] = {
    this.relations.map(_.relation).toSet
  }


  /** Filter set for subject URN matching a given URN.
  *
  * @param u URN to match (either a CTS or CITE2 URN).
  */
  def urn1Match(u: Urn) : CiteRelationSet = {
    val matchingRelations = relations.filter(_.urn1Match(u))
    CiteRelationSet(matchingRelations)
  }

  /** Filter set for object URN matching a given URN.
  *
  * @param u URN to match (either a CTS or CITE2 URN).
  */
  def urn2Match(u: Urn) : CiteRelationSet = {
    val matchingRelations = relations.filter(_.urn2Match(u))
    CiteRelationSet(matchingRelations)
  }


  /** Filter set for triplet matching a given URN
  * (subject or object).
  *
  * @param u URN to match.
  */
  def ~~(u: Urn) : CiteRelationSet = {
    val matchingRelations = relations.filter(_ ~~ u)
    CiteRelationSet(matchingRelations)
  }


  def cex(delimiter: String = "#"): String = {
    val hdr = "#!relations\n"
    val rows = relations.toVector.map(_.cex(delimiter))
    hdr + rows.mkString("\n")
  }

}


/** Factory for creating [[CiteRelationSet]] from Source
* data in CEX format.
*/
object CiteRelationSet extends LogSupport {

  /** Create set from CEX source string.
  *
  * @param cexSrc Source data in CEX format.
  */
  def apply(cexSrc: String, separator: String = "#"): CiteRelationSet = {
    val cex = CexParser(cexSrc)
    val lns = cex.blockString("relations").split("\n").toVector.filter(_.nonEmpty)

    val relations = lns.map(CiteTriple(_))

    /*val colsByLine = lns.map(_.split(separator).toVector.filter(_.size > 0))
    //debug(s"colsbyline: SIZE = ${colsByLine.size}")
    //debug("\t" + colsByLine)

      {

      val triple =  CiteTriple(CiteRelationSet.urnFromString(v(0)), Cite2Urn(v(1)), urnFromString(v(2)))
      triple   } )*/

    if (relations.size > 0) {
      CiteRelationSet(relations.toSet)
    } else {
      CiteRelationSet(Set[CiteTriple]())
    }
  }


  /** Create appropriately typed URN object from String
  * value.
  *
  * @param s String value of either a CtsUrn or Cite2Urn.
  */
  def urnFromString(s: String) : Urn = {
    if (s.startsWith("urn:cite2:")) {
      Cite2Urn(s)
    } else if (s.startsWith("urn:cts:")) {
      CtsUrn(s)
    } else {
      throw(CiteRelationException(s + " is not a URN value."))
    }
  }


}


js/src/main/scala/Main.scala
package edu.holycross.shot.citerelation

import edu.holycross.shot.cite._


/** Verify that JS export works correctly. */
object Main {
  def main(args: Array[String]): Unit = {
    val cexSrc = """
#!relations
urn:cts:greekLit:tlg0012.tlg001.msA:1.title#urn:cite2:cite:dseverbs.r1:illustratedBy#urn:cite2:hmt:vaimg.r1:VA012RN_0013@0.2022,0.211,0.1732,0.0203
"""

//  urn:cts:greekLit:tlg0012.tlg001.msA:1.1#urn:cite2:cite:dseverbs.r1:illustratedBy#urn:cite2:hmt:vaimg.r1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901
  println("Run in JS environment.")
  val il1_1 = CtsUrn("urn:cts:greekLit:tlg0012.tlg001.msA:1.1")
  val img = Cite2Urn("urn:cite2:hmt:vaimg.r1:VA012RN_0013@0.0611,0.2252,0.4675,0.0901")
  val appearsOn = Cite2Urn("urn:cite2:hmt:verbs:appearsOn")


  val triple = CiteTriple(il1_1,appearsOn, img)
  println("Made a CiteTriple " + triple)

  val exc = CiteRelationException("Made up exception")
  println("Made an exception " + exc)

  val tinySet = CiteRelationSet (Set(triple))
  println("Made a set with one relation " + tinySet)

  val relations = CiteRelationSet(cexSrc,"#")
  println("Made a relation set with  " + relations.size + " relation.")


  }
}
